spotifyMini/Models/Analysis.swift

```diff
@@ -65,14 +58,36 @@ class Analysis {
     // MARK: Fetching
 
     func fetchAnalyses() {
-        let ids = self.tracks.map { $0.identifier } as [String]
-        self.spotify.fetchAnalysis(forTrackIDs: ids) { result in
-            self.trackAnalyses = result.value
-            var userInfo: [String:AnyObject]?
-            if let error = result.error as? SpotifyError {
-                userInfo = [AnalysisNotificationErrorUserInfoKey : NSError(spotifyError: error)]
+        let realm = try! Realm()
+        var ids = self.tracks.map { $0.identifier } as [String]
+        let cachedObjects = realm.objects(TrackAnalysis).filter("identifier IN \(ids.realmPredicateFormattedString())")
+        cachedObjects.forEach { self.trackAnalyses.updateValue($0, forKey: $0.identifier) }
+        ids = ids.filter { !self.trackAnalyses.keys.contains($0) }
+
+        if ids.count > 0 {
+            self.spotify.fetchAnalysis(forTrackIDs: ids) { result in
+                if let analyses = result.value {
+                    analyses.forEach { self.trackAnalyses.updateValue($0, forKey: $0.identifier) }
+                    try! realm.write {
+                        realm.add(analyses)
+                    }
+                }
+                NSNotificationCenter.defaultCenter().postNotification(AnalysisFetchedTrackInfoNotification, object: self, optionalError: result.error)
             }
-            NSNotificationCenter.defaultCenter().postNotificationName(AnalysisFetchedTrackInfoNotification, object: self, userInfo: userInfo)
         }
+
+        if cachedObjects.count > 0 {
+            NSNotificationCenter.defaultCenter().postNotification(AnalysisFetchedTrackInfoNotification, object: self)
+        }
+    }
+}
+
+extension _ArrayType where Generator.Element == String {
+    func realmPredicateFormattedString() -> String {
+        var stringSelf = "\(self)"
+        stringSelf = stringSelf.stringByReplacingOccurrencesOfString("[", withString: "{")
```

Comment:
aha....very sneaky here but not a good idea. anyone can override and change the description of an array, including future releases of operating system. componentsJoinedByString should get you pretty far.